### PR TITLE
WIP Simple pool implementation to start integrating

### PR DESCRIPTION
**Pull request by atomirex**

### CODE REVIEW

File: barker/workers/workers.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+				return
+			}
+		}
+	} ()
+
+	return w
+}
+
+func New(numberOfWorkers int) *WorkerPool {
+	queue := make(chan *jobWrapper, numberOfWorkers)
```
Comment (lbouchard-adgear) [inline barker/workers/workers.go:96]: Here are we always gonna want to set the channel buffer size to the number of workers?

File: barker/workers/workers.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func (wp *WorkerPool) Submit(job Job, expiry time.Time) *JobStatus {
+	wrapper := newJobWrapper(job, expiry)
+	wp.queue <- wrapper
+	return wrapper.status()
+}
+
+func (wp *WorkerPool) Close() {
+	// TODO cancel all work in queue
```
Comment (lbouchard-adgear) [inline barker/workers/workers.go:116]: does the TODO still apply?

File: barker/workers/workers.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	close(wp.queue)
+}
```
Comment (lbouchard-adgear) [inline barker/workers/workers.go]: Elegant and simple design :) .

File: barker/workers/workers_test.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
+package workers
+
+import (
+	"testing"
+	"time"
+)
+
+func TestSimple(t *testing.T) {
+	pool := New(4)
+	pool.Close()
```
Comment (lbouchard-adgear) [inline barker/workers/workers_test.go:10]: might be a good idea to check the size of the workers array and, if possible, the buffer size for the channel

File: barker/workers/workers_test.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	switch v := result.(type) {
+	case int:
+		if v != 1 {
+			t.Errorf("Incorrect result from job 2")
+		}
+	default:
+		t.Errorf("Incorrect result type from job execution 2")
+	}
+
+	pool.Close()
```
Comment (lbouchard-adgear) [inline barker/workers/workers_test.go:52]: you might want to put that in a defer statement after creating the pool

File: barker/workers/workers_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	pool := New(1)
+
+	status := pool.Submit(func() (interface{}, error) {
+		return true, nil
+	}, time.Now())
+
+	_, err := status.Wait()
+	if err == nil {
+		t.Errorf("Should have expired and errored")
+	}
```
Comment (lbouchard-adgear) [inline barker/workers/workers_test.go:105]: You might want to check the error message as well, to ensure it is a timeout

File: barker/workers/workers_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+	switch v := result.(type) {
+	case int:
+		if v != 1 {
+			t.Errorf("Incorrect result from job")
+		}
+	default:
+		t.Errorf("Incorrect result type from job execution")
+	}
```
Comment (lbouchard-adgear) [inline barker/workers/workers_test.go:123]: The rest is to check that the pool is not closed?

File: barker/workers/workers.go (hunk 1 above)
Comment (atomirex) [reply barker/workers/workers.go:96]: Yes, you want there to be one more job in the queue ready for each worker, but not so many that we aren't pushing back on clients, that can then load balance with other servers more effectively.

File: barker/workers/workers.go (hunk 2 above)
Comment (atomirex) [reply barker/workers/workers.go:116]: I'm thinking of moving workers into a separate repo, in which case the pools might end up having a lifecycle independent of the app using them, for that the TODO would still apply.

File: barker/workers/workers_test.go (hunk 4 above)
Comment (atomirex) [reply barker/workers/workers_test.go:10]: I'm not sure what you mean here.

File: barker/workers/workers_test.go (hunk 7 above)
Comment (atomirex) [reply barker/workers/workers_test.go:123]: Correct. After one request has timed out it should continue to process other jobs in the queue successfully if those have not expired.

File: barker/server/connection.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+			job, open := <-c.processing
+			if open {
+				result, joberr := job.status.Wait()
+
+				err := binary.Write(c.writer, binary.LittleEndian, job.requestId)
+				if err != nil {
+					return
+				}
+
+				// If error we send back a zero length payload
```
Comment (lbouchard-adgear) [inline barker/server/connection.go:81]: We have to make sure to handle that on barker-client side.

File: barker/server/connection.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	if err != nil || m != n {
+		return
+	}
+
+	err = c.writer.Flush()
+	if err != nil {
+		return
+	}
+
+	go func() {
```
Comment (lbouchard-adgear) [inline barker/server/connection.go:70]: If possible, might be nice to extract this into a separated function.

Comment (lbouchard-adgear) [reply barker/server/connection.go:70]: Does not have to be done in the context of this PR, but eventually, I think that we should have some trace of key errors via logging and/or statsd.  Right now, if I am not mistaking, it is very quiet :) .

File: barker/server/dummyclient.go [hunk 10]
```diff
@@ -1,78 +1,97 @@
 /*  Naive client used for testing purpose
```
Comment (lbouchard-adgear) [inline barker/server/dummyclient.go:1]: Oh thanks for maintaining that :) . That was not necessary.  Hope you did not swear too much lol

Comment (lbouchard-adgear) [reply barker/server/dummyclient.go:1]: Is it only the formatting that is changed?

Comment (lbouchard-adgear) [reply barker/server/dummyclient.go:1]: We should probably make a formatting PR to make sure we apply `go fmt` everywhere ...

File: barker/server/server_test.go [hunk 11]
```diff
@@ -11,7 +11,7 @@ import (
 
 const (
 	PORT = 14790
-	HOST = "0.0.0.0"
+	HOST = "127.0.0.1"
```
Comment (lbouchard-adgear) [inline barker/server/server_test.go]: @juliengrenier  will this be ok with docker?

**Comment by lbouchard-adgear** [comment]
Awesome work :)
Only have very minor comments and some questions.
We should make sure we handle the 0-length payload on barker-client side.  I can help you with that if needed.

File: barker/server/server_test.go (hunk 11 above)
Comment (juliengrenier) [reply barker/server/server_test.go]: most probably via the `-p` option

File: barker/server/dummyclient.go (hunk 10 above)
Comment (atomirex) [reply barker/server/dummyclient.go:1]: I had to make a couple of changes here due to removing writeutils etc. so there's more mention of encoding/binary and so on.

**Comment by atomirex** [comment]
> Awesome work :)
> Only have very minor comments and some questions.
> We should make sure we handle the 0-length payload on barker-client side. I can help you with that if needed.

I wondered if we need to add one method to the handler to retrieve a sort of fixed error payload. (This is because the error format is protocol specific). Then have the zero-length payload interpreted by the client as an unknowable error. This part requires some review, as does the whole requestId stuff.

**Pull request merged**